src/inmanta/data/__init__.py

```diff
@@ -4956,129 +4955,110 @@ async def insert_sets_and_resources(
-        :param connection: The connection to use.
+        :param connection: The connection to use. Must be in a transaction context.
         """
 
         is_partial_update = base_version is not None
 
         updated_resource_sets: set[str | None] = set()
-        resource_data: list[object] = list()
+        resource_data: dict[str, list[object]] = defaultdict(list)
         for r in updated_resources:
-            resource_data.append(
-                {
-                    "resource_id": str(r.resource_id),
-                    "agent": r.agent,
-                    "attributes": r.attributes,
-                    "attribute_hash": util.make_attribute_hash(r.resource_id, r.attributes),
-                    "resource_type": r.resource_type,
-                    "resource_id_value": r.resource_id_value,
-                    "is_undefined": r.is_undefined,
-                    "resource_set": r.resource_set,
-                }
-            )
             updated_resource_sets.add(r.resource_set)
-
-        insert_resource_sets = """
-            -- updated resource sets --
-            WITH resource_set_names(name) AS (
-                SELECT UNNEST($2::text[])
-            )
-            INSERT INTO public.resource_set (environment, id, name)
-            SELECT
-                $1,
-                gen_random_uuid() AS id,
-                rs.name
-            FROM resource_set_names AS rs
-            RETURNING id, name;
-        """
-
-        resource_set_records = await cls._fetch_query(
-            insert_resource_sets, environment, updated_resource_sets, connection=connection
-        )
-
-        # Data to insert into the database
-        resource_set_data = []
-        # Ids of the resource sets we inserted
-        inserted_resource_set_ids = []
-        # Names of the resource sets we do not want to bump in case of a partial compile (outdated and deleted sets)
-        resource_set_names_not_to_bump = set(deleted_resource_sets) if deleted_resource_sets else set()
-        for resource_set in resource_set_records:
-            resource_set_data.append({"id": str(resource_set["id"]), "name": resource_set["name"]})
-            resource_set_names_not_to_bump.add(resource_set["name"])
-            inserted_resource_set_ids.append(str(resource_set["id"]))
-
-        values = [environment, target_version, inserted_resource_set_ids]
+            resource_data["resource_id"].append(str(r.resource_id))
+            resource_data["resource_type"].append(r.resource_type)
+            resource_data["resource_id_value"].append(r.resource_id_value)
+            resource_data["agent"].append(r.agent)
+            resource_data["attributes"].append(r.attributes)
+            resource_data["attribute_hash"].append(util.make_attribute_hash(r.resource_id, r.attributes))
+            resource_data["is_undefined"].append(r.is_undefined)
+            resource_data["resource_set"].append(r.resource_set)
```

Comment:
I would replace the dict with a number of lists 

```python
resource_ids: list[str] = []
resource_type: list[str] = [] 
```

a bit more typed 

(but a detail)

Reply:
On the other hand it makes the `cls._get_value(...)` quite a bit more verbose.

Reply:
get_value is I think a relic from the past. 

In this particular case, it adds nothing, all types used are trivially understood by asyncpg.

And in general, I think we should add [custom type conversion](https://magicstack.github.io/asyncpg/current/usage.html#custom-type-conversions) to asyncpg instead 

and it adds no security, as it just defaults to return value.